source/common/common/callback_impl.h

```diff
@@ -81,5 +104,135 @@ template <typename... CallbackArgs> class CallbackManager {
+
+  /**
+   * @brief Add a callback.
+
+   * @param dispatcher Dispatcher from the same thread as the registered callback. This will be used
+   *                   to schedule the execution of the callback.
+   * @param callback callback to add
+   * @return ThreadSafeCallbackHandlePtr a handle that can be used to remove the callback.
+   */
+  ABSL_MUST_USE_RESULT ThreadSafeCallbackHandlePtr add(Event::Dispatcher& dispatcher,
+                                                       Callback callback) {
+    Thread::LockGuard lock(lock_);
+    auto new_callback = std::make_shared<CallbackHolder>(*this, callback);
+    callbacks_.push_back(
+        CallbackListEntry(std::weak_ptr<CallbackHolder>(new_callback), dispatcher));
+    // Get the list iterator of added callback handle, which will be used to remove itself from
+    // callbacks_ list.
+    new_callback->it_ = (--callbacks_.end());
+    return new_callback;
+  }
+
+  /**
+   * @brief Run all callbacks with a function that returns the input arguments
+   *
+   * NOTE: This code is currently safe if a callback deletes ITSELF from within a callback. It is
+   *       not safe if a callback deletes other callbacks.
+
+   * @param run_with function that is responsible for generating inputs to callbacks. This will be
+   * executed for each callback. Return values must be able to be safely copied across threads and
+   * asynchronously consumed.
+   */
+  void runCallbacksWith(std::function<std::tuple<CallbackArgs...>(void)> run_with) {
+    Thread::LockGuard lock(lock_);
+    for (auto it = callbacks_.cbegin(); it != callbacks_.cend();) {
+      auto& [cb, cb_dispatcher] = *(it++);
+
+      // sanity check cb is valid before attempting to schedule a dispatch
+      if (cb.expired()) {
+        continue;
+      }
+
+      auto args = run_with();
+      cb_dispatcher.post([cb = cb, args = args] {
+        // Once we're running on the thread that scheduled the callback, validate the
+        // callback is still valid and execute
+        std::shared_ptr<CallbackHolder> cb_shared = cb.lock();
+        if (cb_shared != nullptr) {
+          std::apply(cb_shared->cb_, std::move(args));
+        }
+      });
+    }
+  }
+
+  size_t size() const noexcept {
+    Thread::LockGuard lock(lock_);
+    return callbacks_.size();
+  }
+
+private:
+  struct CallbackHolder : public CallbackHandle {
+    CallbackHolder(ThreadSafeCallbackManager& parent, Callback cb)
+        : cb_(cb), parent_dispatcher_(parent.dispatcher_), parent_(parent),
+          still_alive_(parent.still_alive_) {}
+
+    ~CallbackHolder() override {
+      // schedule the removal on the parent thread to avoid cross-thread access or lock contention
+      parent_dispatcher_.post([still_alive = still_alive_, &parent = parent_, it = it_]() mutable {
+        if (!still_alive.expired()) {
+          parent.remove(it);
+        }
+      });
+    }
+
+    Callback cb_;
```

Comment:
I think this data member is unused.

Reply:
This is used when calling the function in `runCallbacksWith`.